theano/sandbox/cuda/tests/test_dnn.py

```diff
@@ -719,6 +719,111 @@ def test_log_softmax2(self):
         utt.assert_allclose(f(inp), f_ref(inp))
 
 
+def test_batchnorm_train_op():
+    if not cuda.dnn.dnn_available():
+        raise SkipTest(cuda.dnn.dnn_available.msg)
+    if cuda.dnn.version() < (5000, 5000):
+        raise SkipTest("batch normalization requires cudnn v5+")
+    utt.seed_rng()
+
+    for mode in ('per-activation', 'spatial'):
+        x, scale, bias = T.tensor4s('x', 'scale', 'bias')
+        eps = 5e-3  # some non-standard value to test if it's used
+
+        # forward pass
+        batchnorm_op = cuda.dnn.GpuDnnBatchNorm(mode=mode, epsilon=eps)
+        out, x_mean, x_invstd = batchnorm_op(x, scale, bias)
+        # reference forward pass
+        if mode == 'per-activation':
+            axes = (0,)
+        elif mode == 'spatial':
+            axes = (0, 2, 3)
+        x_mean2 = x.mean(axis=axes, keepdims=True)
+        x_invstd2 = T.inv(T.sqrt(x.var(axis=axes, keepdims=True) + eps))
+        scale2 = T.addbroadcast(scale, *axes)
+        bias2 = T.addbroadcast(bias, *axes)
+        out2 = (x - x_mean2) * (scale2 * x_invstd2) + bias2
+        # backward pass
+        dy = T.tensor4('dy')
+        grads = T.grad(None, wrt=[x, scale, bias], known_grads={out: dy})
+        # reference backward pass
+        grads2 = T.grad(None, wrt=[x, scale, bias], known_grads={out2: dy})
+        # compile
+        f = theano.function([x, scale, bias, dy],
+                            [out, x_mean, x_invstd, out2, x_mean2, x_invstd2] +
+                            grads + grads2, mode=mode_with_gpu)
+        # run
+        floatX = theano.config.floatX
+        for data_shape in ((10, 20, 30, 40), (4, 3, 2, 1)):
+            param_shape = tuple(1 if d in axes else s
+                                for d, s in enumerate(data_shape))
+            X = 4 + 3 * numpy.random.randn(*data_shape).astype(floatX)
+            Dy = -1 + 2 * numpy.random.randn(*data_shape).astype(floatX)
```

Comment:
Because randn() gives a distribution of zero mean and unit variance, and then batch normalization wouldn't have to change anything. More concretely, it wouldn't allow us to distinguish whether the cuDNN Ops compute variance, standard deviation or the inverses. So I'm adding a bias and coefficient to get a non-zero mean and a non-unit variance.